lib/digital_signal/digital_signal.c

```diff
@@ -72,102 +98,377 @@ uint32_t digital_signal_get_edges_cnt(DigitalSignal* signal) {
+
+    digital_signal_setup_dma(signal);
+    digital_signal_setup_timer();
+    digital_signal_start_timer();
+
+    while(!LL_DMA_IsActiveFlag_TC2(DMA1)) {
+    }
+
+    digital_signal_stop_timer();
+    digital_signal_stop_dma();
+
+    signal->edge_cnt--;
+}
+
+void digital_sequence_alloc_signals(DigitalSequence* sequence, uint32_t size) {
+    sequence->signals_size = size;
+    sequence->signals = malloc(sequence->signals_size * sizeof(DigitalSignal*));
+    sequence->signals_prolonged = malloc(sequence->signals_size * sizeof(bool));
+}
+
+void digital_sequence_alloc_sequence(DigitalSequence* sequence, uint32_t size) {
+    sequence->sequence_used = 0;
+    sequence->sequence_size = size;
+    sequence->sequence = malloc(sequence->sequence_size);
+}
+
+DigitalSequence* digital_sequence_alloc(uint32_t size, const GpioPin* gpio) {
+
+    DigitalSequence* sequence = malloc(sizeof(DigitalSequence));
+
+    sequence->gpio = gpio;
+    sequence->bake = false;
+
+    digital_sequence_alloc_signals(sequence, 32);
+    digital_sequence_alloc_sequence(sequence, size);
+
+    return sequence;
+}
+
+void digital_sequence_free(DigitalSequence* sequence) {
+    furi_assert(sequence);
 
-    while(!LL_DMA_IsActiveFlag_TC2(DMA1))
-        ;
+    free(sequence->signals);
+    free(sequence->sequence);
+    free(sequence);
+}
+
+void digital_sequence_set_signal(DigitalSequence* sequence, uint8_t signal_index, DigitalSignal* signal) {
+    furi_assert(sequence);
+    furi_assert(signal);
+    furi_assert(signal_index < sequence->signals_size);
+
+    sequence->signals[signal_index] = signal;
+    signal->gpio = sequence->gpio;
+    signal->reload_reg_remainder = 0;
+
+    digital_signal_prepare(signal);
+}
+
+void digital_sequence_add(DigitalSequence* sequence, uint8_t signal_index) {
+    furi_assert(sequence);
+    furi_assert(signal_index < sequence->signals_size);
+
+    if(sequence->sequence_used >= sequence->sequence_size) {
+        sequence->sequence_size += 256;
+        sequence->sequence = realloc(sequence->sequence, sequence->sequence_size);
+    }
+
+    sequence->sequence[sequence->sequence_used++] = signal_index;
+}
+
+void digital_signal_update_dma(DigitalSignal* signal) {
+
+    volatile uint32_t dma1_data[] = {
+        /* R6  */ (uint32_t)&(DMA1_Channel1->CCR),
+        /* R7  */ DMA1_Channel1->CCR & ~DMA_CCR_EN,
+        /* R8  */ 2,
+        /* R9  */ (uint32_t)&(signal->gpio->port->BSRR),
+        /* R10 */ (uint32_t)signal->gpio_buff,
+        /* R11 */ DMA1_Channel1->CCR | DMA_CCR_EN };
+
+    volatile uint32_t dma2_data[] = {
+        /* R0 */ (uint32_t)&(DMA1_Channel2->CCR),
+        /* R1 */ DMA1_Channel2->CCR & ~DMA_CCR_EN,
+        /* R2 */ (uint32_t)signal->reload_reg_entries,
+        /* R3 */ (uint32_t)&(TIM2->ARR),
+        /* R4 */ (uint32_t)signal->reload_reg_buff,
+        /* R5 */ DMA1_Channel2->CCR | DMA_CCR_EN };
+
+    
+    /* hurry when setting up next transfer */
+    asm volatile("\t"
```

Comment:
Also thought about using DMAs to automate things, but that would probably occupy all available DMA channels of both DMA.
And even then I am not sure if it would work out.